### PR TITLE
rel-l2-loss: align training loss with surface_rel_l2_pct evaluation metric

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Current training minimises absolute MSE, but the primary evaluation metric on DrivAerML is `surface_rel_l2_pct` (relative L2, in %). This is an objective mismatch: MSE over-weights low-magnitude wake/far-field regions while under-weighting the high-signal-to-noise stagnation and suction peaks that dominate the relative metric. Switching the training objective to a relative L2 formulation (or a mixture) should directly align optimisation with the leaderboard target and improve the metric on all three datasets.

Reference: SMART paper (arxiv 2601.18707) uses relative L2 loss for point-cloud CFD surrogate work and reports DrivAerML-class results. Their relative L2 formulation is adopted here.

## Code Changes

Add the following to `target/icml2026/train.py`:

### 1. Loss functions (add near top of training logic)

```python
def relative_l2_loss(pred, target, eps=1e-8):
    """Per-sample relative L2: ||pred-target||_2 / ||target||_2"""
    diff_norm = (pred - target).norm(dim=-1)
    target_norm = target.norm(dim=-1) + eps
    return (diff_norm / target_norm).mean()

def per_point_relative_l2_loss(pred, target, eps=1e-8):
    """Per-point relative error: |pred-target| / (|target| + eps)"""
    return ((pred - target).abs() / (target.abs() + eps)).mean()

def per_field_relative_l2_loss(pred, target, eps=1e-8):
    """Separate relative L2 per output channel, then mean across channels"""
    # pred/target shape: [..., C]
    diff_norm = (pred - target).norm(dim=0)        # [C]
    target_norm = target.norm(dim=0) + eps          # [C]
    return (diff_norm / target_norm).mean()

def mixed_loss(pred, target, alpha=0.5, eps=1e-8):
    """alpha * MSE + (1 - alpha) * global_rel_l2"""
    mse = F.mse_loss(pred, target)
    rel = relative_l2_loss(pred, target, eps)
    return alpha * mse + (1 - alpha) * rel
```

### 2. CLI flags (add to argparse)

```python
parser.add_argument("--loss", type=str, default="mse",
    choices=["mse", "rel_l2", "per_point_rel_l2", "mixed_rel_l2", "per_field_rel_l2"],
    help="Training loss function")
parser.add_argument("--loss-mix-alpha", type=float, default=0.5,
    help="Alpha weight for mse in mixed_rel_l2 loss (default: 0.5)")
```

### 3. Loss dispatch in training loop

Replace the existing loss call with:
```python
if args.loss == "mse":
    loss = F.mse_loss(pred, target)
elif args.loss == "rel_l2":
    loss = relative_l2_loss(pred, target)
elif args.loss == "per_point_rel_l2":
    loss = per_point_relative_l2_loss(pred, target)
elif args.loss == "per_field_rel_l2":
    loss = per_field_relative_l2_loss(pred, target)
elif args.loss == "mixed_rel_l2":
    loss = mixed_loss(pred, target, alpha=args.loss_mix_alpha)
```

## Baseline

Current best metrics from BASELINE.md (all from PR #2899 / robin, except DrivAerML from PR #2691 / frieren):

| Dataset | Primary metric | Value | PR |
|---|---|---|---|
| DrivAerML | `val_primary/surface_rel_l2_pct` | **4.619%** | #2691 |
| TandemFoil | `val_primary/surface_pressure_mae` | **26.134** | #2899 |
| AirfRANS | `val_primary/surface_mse` | **0.000727** | #2899 |

External DrivAerML target: **<3.71%** (AB-UPT leaderboard).

DrivAerML reproduce command:
```bash
cd target/icml2026 && python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200
```

## Instructions

Use `--wandb_group radford-20260421-relaunch-rel-l2-loss` on **all 8 runs** so they appear together in W&B.

### GPU 0 — DrivAerML: global rel_l2
```bash
cd target/icml2026 && python train.py --dataset drivaerml --loss rel_l2 \
  --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 \
  --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb_group radford-20260421-relaunch-rel-l2-loss
```

### GPU 1 — DrivAerML: per_point_rel_l2
```bash
cd target/icml2026 && python train.py --dataset drivaerml --loss per_point_rel_l2 \
  --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 \
  --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb_group radford-20260421-relaunch-rel-l2-loss
```

### GPU 2 — DrivAerML: mixed_rel_l2 (alpha=0.5)
```bash
cd target/icml2026 && python train.py --dataset drivaerml --loss mixed_rel_l2 --loss-mix-alpha 0.5 \
  --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 \
  --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb_group radford-20260421-relaunch-rel-l2-loss
```

### GPU 3 — DrivAerML: per_field_rel_l2
```bash
cd target/icml2026 && python train.py --dataset drivaerml --loss per_field_rel_l2 \
  --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 \
  --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb_group radford-20260421-relaunch-rel-l2-loss
```

### GPU 4 — TandemFoil: global rel_l2
```bash
cd target/icml2026 && python train.py --dataset tandemfoil --loss rel_l2 \
  --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only \
  --asinh-pressure --residual-prediction --enable-pressure-prior-addition \
  --epochs 999 --ema-decay 0.999 \
  --wandb_group radford-20260421-relaunch-rel-l2-loss
```

### GPU 5 — TandemFoil: per_point_rel_l2
```bash
cd target/icml2026 && python train.py --dataset tandemfoil --loss per_point_rel_l2 \
  --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only \
  --asinh-pressure --residual-prediction --enable-pressure-prior-addition \
  --epochs 999 --ema-decay 0.999 \
  --wandb_group radford-20260421-relaunch-rel-l2-loss
```

### GPU 6 — TandemFoil Paper benchmark: global rel_l2
```bash
cd target/icml2026 && python train.py --dataset tandemfoil_paper --loss rel_l2 \
  --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only \
  --asinh-pressure --residual-prediction --enable-pressure-prior-addition \
  --epochs 999 --ema-decay 0.999 \
  --wandb_group radford-20260421-relaunch-rel-l2-loss
```

### GPU 7 — AirfRANS: global rel_l2
```bash
cd target/icml2026 && python train.py --dataset airfrans --airfrans-task full --loss rel_l2 \
  --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 \
  --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 --ema-decay 0.999 \
  --wandb_group radford-20260421-relaunch-rel-l2-loss
```

See follow-up comment for what to report and suggested next steps.

### CODE REVIEW

**Comment by tcapelle** [comment]
## What to Report

For each GPU run, please report in a results table:

| GPU | Dataset | Loss | val_primary (best) | epoch | W&B run |
|---|---|---|---|---|---|
| 0 | DrivAerML | rel_l2 | | | |
| 1 | DrivAerML | per_point_rel_l2 | | | |
| 2 | DrivAerML | mixed_rel_l2 α=0.5 | | | |
| 3 | DrivAerML | per_field_rel_l2 | | | |
| 4 | TandemFoil | rel_l2 | | | |
| 5 | TandemFoil | per_point_rel_l2 | | | |
| 6 | TandemFoil Paper | rel_l2 | | | |
| 7 | AirfRANS | rel_l2 | | | |

For DrivAerML runs, also report the full surface breakdown at best val checkpoint:
- `val/surface_p_in`, `val/surface_p_oodc`, `val/surface_p_tan`, `val/surface_p_re`

**Key question:** Does aligning the training loss to the evaluation metric (relative L2) yield a lower `surface_rel_l2_pct` than absolute MSE, or does MSE still generalise better to the relative metric?

## Suggested Follow-ups

If `rel_l2` (GPU 0) beats baseline on DrivAerML:
- Try `per_field_rel_l2` with learned per-channel weights
- Try higher lr (1e-3) with relative loss — relative loss has smaller gradient magnitude so may tolerate a larger step

If `mixed_rel_l2` (GPU 2) beats baseline but `rel_l2` does not:
- Sweep alpha: try α=0.2 (less MSE) and α=0.8 (more MSE) to find the optimal mixture

If TandemFoil/AirfRANS relative loss is worse:
- The asinh normalisation on TandemFoil may already stabilise the dynamic range that rel_l2 is meant to address; report whether training loss curves diverge or converge normally

If any run diverges (loss → NaN or explodes):
- Try gradient clipping at 1.0 for that run and restart
